src/graphnet/models/components/layers.py

```
@@ -593,3 +602,393 @@ def forward(
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.residual = residual
+        self.update_edges = update_edges
+        self.batch_norm_momentum = batch_norm_momentum
+        self.batch_norm_runner = batch_norm_runner
+        self.rezero = rezero
+        self.deg_scaler = deg_scaler
+        self.activation = activation()
+
+        self.attention = GritSparseMHA(
+            in_dim=in_dim,
+            out_dim=out_dim // num_heads,
+            num_heads=num_heads,
+            use_bias=attn_bias,
+            dropout=attn_dropout,
+            clamp=attn_clamp,
+            activation=attn_activation,
+            edge_enhance=attn_edge_enhance,
+        )
+
+        self.fc1_x = nn.Linear(out_dim // num_heads * num_heads, out_dim)
+        if enable_edge_transform:
+            self.fc1_e = nn.Linear(out_dim // num_heads * num_heads, out_dim)
+        else:
+            self.fc1_e = nn.Identity()
+
+        if self.deg_scaler:
+            self.deg_coef = nn.Parameter(
+                torch.zeros(1, out_dim // num_heads * num_heads, 2)
+            )
+            nn.init.xavier_normal_(self.deg_coef)
+
+        if norm == nn.LayerNorm:
+            self.norm1_x = norm(out_dim)
+            self.norm1_e = self.norm(out_dim) if norm_edges else nn.Identity()
+        elif norm == nn.BatchNorm1d:
+            self.norm1_x = norm(
+                out_dim,
+                track_running_stats=self.batch_norm_runner,
+                eps=1e-5,
+                momentum=self.batch_norm_momentum,
+            )
+            self.norm1_e = (
+                norm(
+                    out_dim,
+                    track_running_stats=self.batch_norm_runner,
+                    eps=1e-5,
+                    momentum=self.batch_norm_momentum,
+                )
+                if norm_edges
+                else nn.Identity()
+            )
+        else:  # TODO: Maybe just set this to nn.Identity. -PW
```

I think preferred to raise the error if the user passes a non-compatible layer instead of passing warnings and defaulting to identity